### PR TITLE
feat(timeline): Several improvements to MiniGantt

### PR DESCRIPTION
**Pull request by jbrundage**
Fixes GH2585
Fixes GH2584
Fixes GH2583

@GordonSmith please review

### CODE REVIEW

**Comment by GordonSmith** [comment]
Your commit is not signed?

**Comment by GordonSmith** [comment]
Your commit message is all on one line?

**Comment by jbrundage** [comment]
VSCode removes linebreaks from a commit message when amending a commit *sigh*

**Comment by jbrundage** [comment]
@GordonSmith bump

File: packages/common/src/EntityPin.ts [hunk 1]
```diff
@@ -21,34 +21,11 @@ export class EntityPin extends Entity {
...
-        }
-        if (this.annotationIcons().length === 0 || this.annotationOnlyShowOnHover() && !is_hovering) {
-            remove_padding += this.padding();
-        }
-        background_bbox.height -= remove_padding;
+        const background_bbox = this.calcBackgroundBBox(is_hovering, title_bbox, icon_bbox, desc_bbox, _anno_h);
+
+        this._title_widget.display(!(this.titleOnlyShowOnHover() && !is_hovering));
+        this._icon_widget.display(!(this.iconOnlyShowOnHover() && !is_hovering));
+        this._desc_widget.display(!(this.descriptionOnlyShowOnHover() && !is_hovering));
```
Comment (GordonSmith) [inline packages/common/src/EntityPin.ts]: Head wreck:  The following is easier to read IMO (and matches what you have below):
```
!this.descriptionOnlyShowOnHover() || is_hovering
```

File: packages/timeline/src/MiniGantt.ts [hunk 2]
```diff
@@ -225,34 +216,38 @@ export class MiniGantt extends SVGWidget {
...
-        this.gUpper
-            .attr("class", "gUpper")
-            .attr("transform", `translate(0,${gUpperTransY})`)
-            ;
-        this.gLower
-            .attr("class", "gLower")
-            .attr("transform", `translate(0,${upperHeight})`)
-            ;
+        const upperContentHeight = this.updateEntityPins(events);
+        const upperAxisHeight = 28; // this.gUpperAxis.node().getBBox().height;
```
Comment (GordonSmith) [inline packages/timeline/src/MiniGantt.ts]: The upperAxis might have labels which are staggered and its height may well be larger than 28?

File: packages/timeline/src/MiniGantt.ts [hunk 3]
```diff
@@ -225,34 +216,38 @@ export class MiniGantt extends SVGWidget {
...
+        const lowerHeight = height - upperContentHeight;
+        console.log("upperHeight", upperContentHeight);
```
Comment (GordonSmith) [inline packages/timeline/src/MiniGantt.ts]: Remove console.

File: packages/timeline/src/MiniGantt.ts [hunk 4]
```diff
@@ -440,3 +461,49 @@ export class MiniGantt extends SVGWidget {
...
+    eventBorderColorColumn(_: string): this;
+    eventBackgroundColorColumn(): string;
+    eventBackgroundColorColumn(_: string): this;
+}
+
+MiniGantt.prototype.publish("timePattern", "%Y-%m-%d", "string", "timePattern");
+MiniGantt.prototype.publish("tickFormat", "%Y-%m-%d", "string", "tickFormat");
+MiniGantt.prototype.publish("tooltipTimeFormat", "%Y-%m-%d", "string", "tooltipTimeFormat");
+MiniGantt.prototype.publish("overlapTolerence", 2, "number", "overlapTolerence");
+MiniGantt.prototype.publish("orientation", "horizontal", "set", "orientation", ["horizontal", "vertical"]);
```
Comment (GordonSmith) [inline packages/timeline/src/MiniGantt.ts]: Have you retested with orientatio=vertical?

File: packages/timeline/src/MiniGantt.ts (hunk 2 above)
Comment (jbrundage) [reply packages/timeline/src/MiniGantt.ts]: I'm thinking that dynamically determining the height of the upper axis at that point would require the Axis synchronously calculating its height. Any suggestions on how to accomplish that?

**Comment by GordonSmith** [comment]
@jchambers-ln Can you switch this code to use EntityRect instead of EntityCard?

**Comment by jbrundage** [comment]
@GordonSmith this is ready to be reviewed

File: packages/common/src/EntityPin.ts [hunk 5]
```diff
@@ -17,38 +17,16 @@ export class EntityPin extends Entity {
         this._icon_widget.render();
         this._desc_widget.render();
         this._title_widget.render();
+        console.log("this.descriptionOnlyShowOnHover()", this.descriptionOnlyShowOnHover());
```
Comment (GordonSmith) [inline packages/common/src/EntityPin.ts]: Remove console log

File: packages/common/src/EntityPin.ts [hunk 6]
```diff
@@ -72,6 +50,42 @@ export class EntityPin extends Entity {
         this.moveAnnotations(background_bbox.width / 2, anno_y);
 
     }
+    calcHeight() {
+        const is_hovering = true;
+        console.log("this.descriptionOnlyShowOnHover()", this.descriptionOnlyShowOnHover());
```
Comment (GordonSmith) [inline packages/common/src/EntityPin.ts]: Remove console.log

File: packages/timeline/src/MiniGantt.ts (hunk 2 above)
Comment (GordonSmith) [reply packages/timeline/src/MiniGantt.ts]: You can ask it for its height once its been rendered

**Comment by GordonSmith** [comment]
@jbrundage Travis error (legitimate)

**Comment by jbrundage** [comment]
@GordonSmith this code change fixes the top padding issue when displaying only ranges as well as the tiered event height enhancement for OTTO

**Pull request merged**